### ISSUE TITLE
[Serve] f.options().deploy() cannot recover from checkpoints upon ray controller failure

### ISSUE THREAD

**Issue by jiaodong** [ray-project/ray#17569]
### What is the problem?

In case of multiple deploy() calls of the same deployment, we might not have correct checkpoint in some cases under controller actor failure.

### Reproduction (REQUIRED)

In ray serve tests, enable controller failure test case and run 

`pytest test_failure.py::test_controller_failure`

Test will consistently fail, but passes after commenting out the first ray.kill

    # ray.kill(serve.api._global_client._controller, no_restart=False)
    function3.deploy()
    ray.kill(serve.api._global_client._controller, no_restart=False)

In other words, it will only pass if deploy() gets called first, which suggesting previous f.options().deploy() is not checkpointed correctly without controller fault tolerance.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by edoakes** [comment]
Interestingly, this minimal version of the behavior passes consistently for me locally:
```python
 24 def test_controller_failure(serve_instance):
 25     @serve.deployment(name="controller_failure")
 26     def function(_):
 27         return "hello1"
 28
 29     function.deploy()
 30
 31     response = request_with_retries("/controller_failure/", timeout=30)
 32     assert response.text == "hello1"
 33
 34     def function2(_):
 35         return "hello2"
 36
 37     function.options(func_or_class=function2).deploy()
 38
 39     ray.kill(serve.api._global_client._controller, no_restart=False)
 40
 41     for _ in range(10):
 42         response = request_with_retries("/controller_failure/", timeout=30)
 43         assert response.text == "hello2"
```

**Comment by edoakes** [comment]
@jiaodong I believe this is fixed now that the test is enabled, right?
